services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/NonLinearMultipartDownloaderSubscriber.java

```diff
@@ -0,0 +1,399 @@
+
+    private final AtomicBoolean completedExceptionally = new AtomicBoolean(false);
+
+    public NonLinearMultipartDownloaderSubscriber(S3AsyncClient s3,
+                                                  GetObjectRequest getObjectRequest,
+                                                  CompletableFuture<GetObjectResponse> resultFuture,
+                                                  int maxInFLight) {
+        this.s3 = s3;
+        this.getObjectRequest = getObjectRequest;
+        this.resultFuture = resultFuture;
+        this.maxInFlightParts = maxInFLight;
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        if (this.subscription != null) {
+            s.cancel();
+            return;
+        }
+        this.subscription = s;
+        request(1);
+    }
+
+    private void request(int amount) {
+        outstandingDemand.addAndGet(amount);
+        transformersRequested.addAndGet(amount);
+        subscription.request(amount);
+    }
+
+    @Override
+    public void onNext(AsyncResponseTransformer<GetObjectResponse, GetObjectResponse> asyncResponseTransformer) {
+        outstandingDemand.decrementAndGet();
+        log.trace(() -> "On Next - Total in flight parts: " + inFlightRequests.size()
+                        + " - Demand : " + outstandingDemand.get()
+                        + " - Total completed parts: " + completedParts
+                        + " - Total transformers requested: " + transformersRequested.get()
+                        + " - Total pending transformers: " + pendingTransformers.size()
+                        + " - Current in flight requests: " + inFlightRequests.keySet());
+        if (asyncResponseTransformer == null) {
+            subscription.cancel();
+            throw new NullPointerException("onNext must not be called with null asyncResponseTransformer");
+        }
+        executeRequestOrAddToPending(asyncResponseTransformer);
+    }
+
+    private void executeRequestOrAddToPending(
+        AsyncResponseTransformer<GetObjectResponse, GetObjectResponse> asyncResponseTransformer) {
+        if (handleFirstRequest(asyncResponseTransformer)) {
+            return;
+        }
+
+        if (inFlightRequests.size() >= maxInFlightParts) {
+            pendingTransformers.offer(asyncResponseTransformer);
+            return;
+        }
+
+        sendNextRequest(asyncResponseTransformer);
+        requestMoreIfNeeded();
+    }
+
+    // Handle first request.
+    // We need to wait for the first request to finish so we know if it is aa multipart object or not.
+    // While we don't know yet, additional onNext signal receives are stored in pendingTransformers.
+    private boolean handleFirstRequest(AsyncResponseTransformer<GetObjectResponse, GetObjectResponse> asyncResponseTransformer) {
+        if (completedParts.get() != 0) {
+            return false;
+        }
+
+        firstPartLock.lock();
```

Comment:
Is there any reason we need this? We only request more in whenComplete future of the first part, right?

MPU has similar logic where we need to wait for creatMPU to finish, can we use similar logic here? https://github.com/aws/aws-sdk-java-v2/blob/master/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithUnknownContentLengthHelper.java#L191

Reply:
outdated after class refactoring